src/views/shared/amount-input/amount-input.jsx

```diff
@@ -0,0 +1,145 @@
+
+function AmountInput (Component) {
+  return function (props) {
+    const { transactionType, account, l2Fee, fiatExchangeRates, preferredCurrency } = props
+    const [gasPrice, setGasPrice] = React.useState(BigNumber.from(0))
+    const [value, setValue] = React.useState('')
+    const [amount, setAmount] = React.useState({ tokens: BigNumber.from(0), fiat: 0 })
+    const [showInFiat, setShowInFiat] = React.useState(false)
+    const [isAmountNegative, setIsAmountNegative] = React.useState(undefined)
+    const [isAmountMoreThanFunds, setIsAmountMoreThanFunds] = React.useState(undefined)
+    const [isAmountCompressedInvalid, setIsAmountCompressedInvalid] = React.useState(undefined)
+
+    React.useEffect(() => {
+      getProvider().getGasPrice().then((gasPrice) => setGasPrice(gasPrice))
+    }, [])
+
+    React.useEffect(() => {
+      if (props.onChange) {
+        const isInvalid = (() => {
+          if (isAmountNegative === undefined && isAmountMoreThanFunds === undefined && isAmountCompressedInvalid === undefined) {
+            return undefined
+          }
+
+          return isAmountNegative || isAmountMoreThanFunds || isAmountCompressedInvalid
+        })()
+
+        props.onChange({
+          amount,
+          showInFiat,
+          isInvalid
+        })
+      }
+    }, [amount, showInFiat, isAmountMoreThanFunds, isAmountCompressedInvalid])
+
+    function convertAmountToFiat (tokensAmount) {
+      const fixedTokenAmount = getFixedTokenAmount(tokensAmount.toString(), account.token.decimals)
+
+      return getTokenAmountInPreferredCurrency(
+        fixedTokenAmount,
+        account.token.USD,
+        preferredCurrency,
+        fiatExchangeRates
+      ).toFixed(2)
+    }
+
+    function convertAmountToTokens (fiatAmount) {
+      const tokensAmount = fiatAmount / account.token.USD
+
+      return parseUnits(tokensAmount.toFixed(account.token.decimals))
+    }
+
+    function checkAmountValidity (newAmount) {
+      const fee = getTransactionFee(transactionType, newAmount, account.token, l2Fee, gasPrice)
+      const newAmountWithFee = newAmount.add(fee)
+
+      setIsAmountNegative(newAmountWithFee.lte(BigNumber.from(0)))
+      setIsAmountMoreThanFunds(newAmountWithFee.gt(BigNumber.from(account.balance)))
+
+      if (transactionType !== TxType.Deposit && transactionType !== TxType.ForceExit) {
+        setIsAmountCompressedInvalid(isTransactionAmountCompressedValid(newAmount) === false)
+      }
+    }
+
+    function handleInputChange (event) {
+      if (INPUT_REGEX.test(event.target.value)) {
+        if (showInFiat) {
+          const newAmountInFiat = Number(event.target.value)
+          const newAmountInTokens = convertAmountToTokens(newAmountInFiat)
+          const fixedAmountInTokens = fixTransactionAmount(newAmountInTokens)
+
+          setAmount({ tokens: fixedAmountInTokens, fiat: newAmountInFiat.toFixed(2) })
+          checkAmountValidity(fixedAmountInTokens)
+          setValue(event.target.value)
```

Comment:
Why do we need this if we never modify it?

Reply:
As this is a `controlled` input we need to store the value to pass it down to the html `input`. We need this to prevent user from input invalid characters, which is not possible with `uncontrolled` inputs.